lib/global_uid/allocator.rb

```diff
@@ -0,0 +1,66 @@
+module GlobalUid
+  class Allocator
+    attr_reader :allocations, :max_window_size, :incrementing_by, :connection
+
+    def initialize(max_window_size: 5, incrementing_by:, connection:)
+      @allocations = []
+      @max_window_size = max_window_size
+      @incrementing_by = incrementing_by
+      @connection = connection
+      validate_connection_increment
+    end
+
+    def allocate_one(table)
+      identifier = connection.insert("REPLACE INTO #{table} (stub) VALUES ('a')")
+      allocate(identifier)
+    end
+
+    def allocate_many(table, count:)
+      increment_by = validate_connection_increment
+
+      start_id = connection.insert("REPLACE INTO #{table} (stub) VALUES " + (["('a')"] * count).join(','))
+      identifiers = start_id.step(start_id + (count - 1) * increment_by, increment_by).to_a
+      identifiers.each { |identifier| allocate(identifier) }
+      identifiers
+    end
+
+    private
+
+    def allocate(identifier)
+      allocations.shift if allocations.size >= max_window_size
+      allocations << identifier
+
+      if !valid_allocation?
+        db_increment = connection.select_value("SELECT @@auto_increment_increment")
+        message = "Configured: '#{incrementing_by}', Found: '#{db_increment}' on '#{connection.current_database}'. Recently allocated IDs: #{allocations}"
+        alert(InvalidIncrementException.new(message))
+      end
+
+      identifier
+    end
+
+    def valid_allocation?
+      allocations[1..-1].all? do |identifier|
+        (identifier - allocations[0]) % incrementing_by == 0
+      end
+    end
+
```

Comment:
An additional check I propose we do is to see if these ids are increasing order only. If an id goes backwards, we should alert as wel.